sds/src/secondary_validation/finnish_hetu_checksum.rs

```diff
@@ -0,0 +1,110 @@
+use crate::secondary_validation::Validator;
+
+pub struct FinnishHetuChecksum;
+
+const HETU_LENGTH: usize = 11;
+const CONTROL_CHARS: &str = "0123456789ABCDEFHJKLMNPRSTUVWXY";
+
+impl Validator for FinnishHetuChecksum {
+    fn is_valid_match(&self, regex_match: &str) -> bool {
+        /*
+         * Finnish personal identification numbers (HETU) format:
+         * DDMMYYCZZZQ where:
+         * - DDMMYY is the date of birth
+         * - C is the century marker (-, +, A-F, U-Y)
+         * - ZZZ is the individual number (002-899 for regular, 900-999 for temporary)
+         * - Q is the control character
+         */
+        if regex_match.len() != HETU_LENGTH {
+            return false;
+        }
+
+        // Remove the optional spaces for processing
+        let processed_match = regex_match.replace(' ', "");
+        if processed_match.len() != HETU_LENGTH {
+            return false;
+        }
+
+        // Split the components
+        let date_part = &processed_match[..6];
+        let century_marker = processed_match.chars().nth(6).unwrap();
+        let individual_number = &processed_match[7..10];
+        let control_char = processed_match.chars().last().unwrap();
+
+        // Validate date part (DDMMYY)
+        if !date_part.chars().all(|c| c.is_ascii_digit()) {
+            return false;
+        }
+
+        // Validate century marker, the law (https://vm.fi/paatos?decisionId=0900908f807c5f3c) mention the following syntax
+        // For those born on January 1, 2000, or later, the separator used is the letter A, B, C, D, E, or F; for those born in the 1900s, a hyphen (-) or the letter Y, X, W, V, or U is used; and for those born in the 1800s, a plus sign (+) is used.
+        if !matches!(century_marker, '-' | '+' | 'A'..='F' | 'U'..='Y') {
+            return false;
+        }
+
+        // Validate individual number
+        if !individual_number.chars().all(|c| c.is_ascii_digit()) {
+            return false;
+        }
+
+        // Validate control character
+        if !CONTROL_CHARS.contains(control_char.to_ascii_uppercase()) {
+            return false;
+        }
```

Comment:
The regex should already check these characters, the checksum can only focus on validating the control character.